### PR TITLE
Add Header and Method support for HTTP checks

### LINKED ISSUE #2924: Enable ability to send username/password for Basic Auth in service check stanza
Nomad v0.6

Request for ability to pass username/password for Basic Auth in the service check stanza.  As example, I have below where it checks the cluster endpoints for ElasticSearch:
```
      service {
        name = "escluster"
        tags = [
          "search",
          "cluster"
        ]
        port = "eshttp"
        check {
          name = "available"
          type = "tcp"
          interval = "10s"
          timeout = "2s"
        }
        check {
          name = "ready"
          type = "http"
          port = "eshttp"
          path = "/_cluster/health?wait_for_status=yellow"
          interval = "30s"
          timeout = "10s"
        }
      }
```

As cannot authenticate; will hit error below:
```
HTTP GET http://10.0.42.4:47553/_cluster/health?wait_for_status=yellow: 401 Unauthorized Output: {"error":{"root_cause":[{"type":"security_exception","reason":"missing authentication token for REST request [/_cluster/health?wait_for_status=yellow]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}}],"type":"security_exception","reason":"missing authentication token for REST request [/_cluster/health?wait_for_status=yellow]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}},"status":401}
```

Thanks for any possible guidance.  Thanks!

### PR DESCRIPTION
**Pull request by schmichael**
Fixes #2924 

@dadgar Please check my work in diff.go. I haven't had to edit the diff logic before.

I modeled our header field after Consul's which is modeled after the [Go stdlib](https://golang.org/pkg/net/http/#Header) which ... is kind of awkward for people not used to it for 2 reasons:

1. Header (singluar) when people probably expect _Headers_ (plural)
2. `map[string][]string` instead of `[]string` (like env vars)

It's unfortunate that it's kind of an awkward API, but I default to consistency above trying to make a subjective UX tweak.

### CODE REVIEW

File: command/agent/consul/unit_test.go [hunk 1]
```diff
@@ -1352,3 +1354,45 @@ func TestIsNomadService(t *testing.T) {
...
+	checkID := check.Hash(serviceID)
+	host := "localhost"
+	port := 41111
+
+	expected := &api.AgentCheckRegistration{
+		ID:        checkID,
+		Name:      "name",
+		ServiceID: serviceID,
+	}
+	expected.Timeout = "0s"
```
Comment (dadgar) [inline command/agent/consul/unit_test.go]: Any reason these aren't inline with struct creation?

File: jobspec/parse.go [hunk 2]
```diff
@@ -971,6 +973,37 @@ func parseChecks(service *api.Service, checkObjs *ast.ObjectList) error {
 		if err := hcl.DecodeObject(&cm, co.Val); err != nil {
 			return err
 		}
+
+		// HCL allows repeating stanzas so merge 'header' into a single
+		// map[string][]string.
+		if headerI, ok := cm["header"]; ok {
+			headerRaw, ok := headerI.([]map[string]interface{})
+			if !ok {
+				return fmt.Errorf("check -> header -> expected a []map[string]interface{} but found %T", headerI)
```
Comment (dadgar) [inline jobspec/parse.go]: Should we show interface{} or []string since that is the only thing that is parsable.

Comment (dadgar) [inline jobspec/parse.go]: Should we show interface{} or []string since that is the only thing that is parsable.

File: jobspec/parse_test.go [hunk 3]
```diff
@@ -449,9 +449,14 @@ func TestParse(t *testing.T) {
 											{
 												Name:          "check-name",
 												Type:          "http",
+												Path:          "/",
 												Interval:      10 * time.Second,
 												Timeout:       2 * time.Second,
 												InitialStatus: capi.HealthPassing,
+												Method:        "POST",
```
Comment (dadgar) [inline jobspec/parse_test.go]: Maybe a failing case? With numbers or something.

File: nomad/structs/diff.go [hunk 4]
```diff
@@ -592,6 +592,23 @@ func serviceCheckDiff(old, new *ServiceCheck, contextual bool) *ObjectDiff {
 
 	// Diff the primitive fields.
 	diff.Fields = fieldDiffs(oldPrimitiveFlat, newPrimitiveFlat, contextual)
+
+	// Diff Header
+	headerDiff := &ObjectDiff{Type: DiffTypeNone, Name: "Header"}
```
Comment (dadgar) [inline nomad/structs/diff.go]: Would you mind just popping it into a method?

**Comment by schmichael** [comment]
Travis failure is unrelated.

**Pull request merged**